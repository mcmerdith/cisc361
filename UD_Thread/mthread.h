/* 
 * thread library function prototypes
 */
void t_create(void(*function)(int), int thread_id, int priority);
void t_yield(void);
void t_init(void);
void t_terminate(void);
void t_shutdown(void);
