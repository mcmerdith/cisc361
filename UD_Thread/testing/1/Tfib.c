/*
 * Multithreaded Fibonacci Generator
 * By: Kyle Sullivan
 * University of Delaware - CISC 361
 *
 * This test case demonstrates how to generate fibonacci numbers using
 * multithreading and dynamic programming
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include "mthread.h"

// this lock is used to ensure mutual exclusion
// when accessing the fibonacci array across
// multiple threads
pthread_rwlock_t rwlock = PTHREAD_RWLOCK_INITIALIZER;

// global array to store fibonacci results
long long fib_array[10000];
// note NUM_THREADS and N have to be divisible
int NUM_THREADS = 4;
int N = 92; // max for 64 bits

/*
 * fib
 * - fib(n) returns the nth number of the fibonacci sequence
 *   (note the answer is found using dynamic programming,
 *    this means that previous answers are stored in an array
 *    to speed up the calculation time of the nth + 1 fib number)
 */
long long fib(int n)
{

   // first cover the base cases, no lock
   // needed since the first two elements are given
   if (n == 0)
      return fib_array[0];

   if (n == 1)
      return fib_array[1];

   // retrieve the n - 2 fibonacci number,
   // if it is 0 (aka not previously calculated)
   // then lets calculate it
   pthread_rwlock_rdlock(&rwlock);
   long long tmp1 = fib_array[n - 2];
   pthread_rwlock_unlock(&rwlock);
   if (tmp1 == 0)
   {
      fib(n - 2);
   }

   // retrieve the n - 1 fibonacci number,
   // if it is 0 (aka not previously calculated)
   // then lets calculate it
   pthread_rwlock_rdlock(&rwlock);
   long long tmp2 = fib_array[n - 2];
   pthread_rwlock_unlock(&rwlock);
   if (tmp2 == 0)
   {
      fib(n - 1);
   }

   // we have the n - 2 and n - 1 fibonacci number,
   // now we can calculate the nth fibonacci number
   // when we finish calculating it, store it in the array
   // so we don't don't have to calculate it again later
   long long rval;
   pthread_rwlock_wrlock(&rwlock);
   fib_array[n] = fib_array[n - 1] + fib_array[n - 2];
   rval = fib_array[n];
   pthread_rwlock_unlock(&rwlock);

   // return the nth fiboacci number
   return rval;
}

// generates a thread which calculates fiboacci numbers
void fib_thread(int thr_id)
{

   int n;
   long long fibnum;
   // int n = thr_id;

   // loops through and calculates the fiboacci numbers
   // for the respective thread based on the total number
   // of threads, so if there are 4 threads and this is thread 0,
   // it will find fib(0), fib(4), fib(8) . . . until fib(N)
   for (n = thr_id; n < N; n = n + NUM_THREADS)
   {
      fibnum = fib(n);
      printf("fib[%d] = %lld (thread #%d)\n", n, fibnum, thr_id);
      t_yield();
   }

   printf("Thread %d is done...\n", thr_id);
   t_terminate();
}

/*
 * USAGE:
 * ./multi_fib <N> <num_threads>
 * - generates from 0 to Nth fibonacci number using
 *   the number of threads specified
 *
 * - default, N = 92 (max for 64 bits), and 4 threads
 *
 * NOTE: N and NUM_THREADS need to be divisible
 */
int main(int argc, char *argv[])
{

   // parse the command line for N and NUM_THREADS
   if (argc == 3)
   {
      N = atoi(argv[1]);
      NUM_THREADS = atoi(argv[2]);

      if (N < 0 || N > 92)
      {
         printf("N must be between 0 and 92.\n");
         return 1;
      }

      if (NUM_THREADS < 1)
      {
         printf("NUM_THREADS must be at least 1 thread or more.\n");
         return 1;
      }

      if (N % NUM_THREADS != 0)
      {
         printf("N and NUM_THREADS must be divisible\n");
         return 1;
      }
   }
   printf("Generating the fibonacci sequence from 0 to %d, using %d threads.\n", N, NUM_THREADS);

   // Initialize the beginning of the fibonacci array
   fib_array[0] = 0;
   fib_array[1] = 1;

   // thread library, main thread starts
   t_init();

   // create the threads base on NUM_THREADS
   int i;
   for (i = 0; i < NUM_THREADS; i++)
   {
      t_create(fib_thread, i, 1);
      printf("created fib_thread %d\n", i);
   }

   // go longer than all other threads (to show when t_yield is called with empty ready queue)
   for (i = 0; i < N + 1; i++)
   {

      t_yield();
   }

   // all done
   printf("Begin shutdown...\n");
   t_shutdown();
   printf("Done with shutdown...\n");

   return 0;
}
