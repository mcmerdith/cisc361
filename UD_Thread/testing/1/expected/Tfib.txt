Generating the fibonacci sequence from 0 to 92, using 4 threads.
created fib_thread 0
created fib_thread 1
created fib_thread 2
created fib_thread 3
fib[0] = 0 (thread #0)
fib[1] = 1 (thread #1)
fib[2] = 1 (thread #2)
fib[3] = 2 (thread #3)
fib[4] = 3 (thread #0)
fib[5] = 5 (thread #1)
fib[6] = 8 (thread #2)
fib[7] = 13 (thread #3)
fib[8] = 21 (thread #0)
fib[9] = 34 (thread #1)
fib[10] = 55 (thread #2)
fib[11] = 89 (thread #3)
fib[12] = 144 (thread #0)
fib[13] = 233 (thread #1)
fib[14] = 377 (thread #2)
fib[15] = 610 (thread #3)
fib[16] = 987 (thread #0)
fib[17] = 1597 (thread #1)
fib[18] = 2584 (thread #2)
fib[19] = 4181 (thread #3)
fib[20] = 6765 (thread #0)
fib[21] = 10946 (thread #1)
fib[22] = 17711 (thread #2)
fib[23] = 28657 (thread #3)
fib[24] = 46368 (thread #0)
fib[25] = 75025 (thread #1)
fib[26] = 121393 (thread #2)
fib[27] = 196418 (thread #3)
fib[28] = 317811 (thread #0)
fib[29] = 514229 (thread #1)
fib[30] = 832040 (thread #2)
fib[31] = 1346269 (thread #3)
fib[32] = 2178309 (thread #0)
fib[33] = 3524578 (thread #1)
fib[34] = 5702887 (thread #2)
fib[35] = 9227465 (thread #3)
fib[36] = 14930352 (thread #0)
fib[37] = 24157817 (thread #1)
fib[38] = 39088169 (thread #2)
fib[39] = 63245986 (thread #3)
fib[40] = 102334155 (thread #0)
fib[41] = 165580141 (thread #1)
fib[42] = 267914296 (thread #2)
fib[43] = 433494437 (thread #3)
fib[44] = 701408733 (thread #0)
fib[45] = 1134903170 (thread #1)
fib[46] = 1836311903 (thread #2)
fib[47] = 2971215073 (thread #3)
fib[48] = 4807526976 (thread #0)
fib[49] = 7778742049 (thread #1)
fib[50] = 12586269025 (thread #2)
fib[51] = 20365011074 (thread #3)
fib[52] = 32951280099 (thread #0)
fib[53] = 53316291173 (thread #1)
fib[54] = 86267571272 (thread #2)
fib[55] = 139583862445 (thread #3)
fib[56] = 225851433717 (thread #0)
fib[57] = 365435296162 (thread #1)
fib[58] = 591286729879 (thread #2)
fib[59] = 956722026041 (thread #3)
fib[60] = 1548008755920 (thread #0)
fib[61] = 2504730781961 (thread #1)
fib[62] = 4052739537881 (thread #2)
fib[63] = 6557470319842 (thread #3)
fib[64] = 10610209857723 (thread #0)
fib[65] = 17167680177565 (thread #1)
fib[66] = 27777890035288 (thread #2)
fib[67] = 44945570212853 (thread #3)
fib[68] = 72723460248141 (thread #0)
fib[69] = 117669030460994 (thread #1)
fib[70] = 190392490709135 (thread #2)
fib[71] = 308061521170129 (thread #3)
fib[72] = 498454011879264 (thread #0)
fib[73] = 806515533049393 (thread #1)
fib[74] = 1304969544928657 (thread #2)
fib[75] = 2111485077978050 (thread #3)
fib[76] = 3416454622906707 (thread #0)
fib[77] = 5527939700884757 (thread #1)
fib[78] = 8944394323791464 (thread #2)
fib[79] = 14472334024676221 (thread #3)
fib[80] = 23416728348467685 (thread #0)
fib[81] = 37889062373143906 (thread #1)
fib[82] = 61305790721611591 (thread #2)
fib[83] = 99194853094755497 (thread #3)
fib[84] = 160500643816367088 (thread #0)
fib[85] = 259695496911122585 (thread #1)
fib[86] = 420196140727489673 (thread #2)
fib[87] = 679891637638612258 (thread #3)
fib[88] = 1100087778366101931 (thread #0)
fib[89] = 1779979416004714189 (thread #1)
fib[90] = 2880067194370816120 (thread #2)
fib[91] = 4660046610375530309 (thread #3)
Thread 0 is done...
Thread 1 is done...
Thread 2 is done...
Thread 3 is done...
Begin shutdown...
Done with shutdown...
